vision.py

```diff
@@ -154,43 +132,51 @@ def find_power_port(self, frame: np.ndarray) -> tuple:
                     acceptable_cnts.append(current_contour[1])
 
-            power_port_contour = max(acceptable_cnts, key=lambda x: cv2.contourArea(x))
-            power_port_points = self.find_polygon(power_port_contour)
+            if acceptable_cnts:
+                power_port_contour = max(
+                    acceptable_cnts, key=lambda x: cv2.contourArea(x)
+                )
+                power_port_points = get_corners_from_contour(power_port_contour)
             # x, y, w, h = cv2.boundingRect(power_port_contour)
-            return power_port_contour, power_port_points
+                return power_port_points
+            else:
+                return None
         else:
             return None
 
-    def create_annotated_display(self, frame: np.ndarray, points: np.ndarray, printing=False):
+    def create_annotated_display(
+        self, frame: np.ndarray, points: np.ndarray, printing=False
+    ):
         for i in range(len(points)):
             cv2.circle(frame, (points[i][0][0], points[i][0][1]), 5, (0, 255, 0))
         if printing == True:
             print(points)
 
+    def get_vertical_angle(self, p: int):
+        """Gets angle of point p above the horizontal.
+        Parameter p should have 0 at the bottom of the frame and FRAME_HEIGHT at the top. """
+        return math.atan2(p - FRAME_HEIGHT, FY)
+
     # get_angle and get_distance will be replaced with solve pnp eventually
-    def get_angle(self, X:float) -> float:
+    def get_horizontal_angle(self, X: float) -> float:
         return (
             ((X / FRAME_WIDTH) - 0.5) * MAX_FOV_WIDTH * self.zoom / 100
         )  # 33.18 degrees #gets the angle
 
-    def get_distance(self, contour: np.ndarray, angle: float) -> float:
-        box = cv2.boundingRect(contour)
-        width = box[2]
-        distance = (
-            PORT_DIMENTIONS[0]
-            / math.tan((width / FRAME_WIDTH) * (MAX_FOV_WIDTH / 2))
-            * (self.zoom / 100)
-        )  # the current method this uses is not mathmetically correct, the correct method would use the law of cosines
-        # this just uses a tan and then tries to correct itself
-        distance -= angle * 1.9
-        distance *= 0.6
-        return distance
-
-    def get_middle(self, contour: np.ndarray) -> tuple:
-        # https://www.pyimagesearch.com/2016/02/01/opencv-center-of-contour/
+    def get_distance(self, Y: float) -> float:
+        target_angle = self.get_vertical_angle(Y)
+        # print(f"Total Angle: {math.degrees(target_angle + GROUND_ANGLE)}")
+        return (TARGET_HEIGHT - CAMERA_HEIGHT) / math.tan(GROUND_ANGLE + target_angle)
+
+    def get_middles(self, contour: np.ndarray) -> tuple:
+        """ Use the cv2 moments to find the centre x and y of the contour. """
         M = cv2.moments(contour)
+        if M["m00"] != 0:
         cX = int(M["m10"] / M["m00"])
-        cY = int(M["m01"] / M["m00"])
+        else:
+            cX = 160
+        # cY = int(M["m01"] / M["m00"])
```

Comment:
Does this need to be here anymore?

Reply:
Consider replacing this with a comment to the effect that we are after the center of the hexagon, not of the shape we see.

Reply:
We use the bottom of the hexagon.